src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java

```diff
@@ -15,152 +15,191 @@
+    log.debug("Flushing data to Elasticsearch with the following offsets: {}", offsets);
+    try {
+      client.flush();
+    } catch (IllegalStateException e) {
+      log.debug("Tried to flush data to Elasticsearch, but BulkProcessor is already closed.");
     }
   }
 
   @Override
-  public void put(Collection<SinkRecord> records) throws ConnectException {
-    log.debug("Putting {} records to Elasticsearch", records.size());
-    writer.write(records);
+  public void stop() {
+    log.debug("Stopping Elasticsearch client.");
+    client.close();
+    log.debug("Stopped Elasticsearch client.");
   }
 
   @Override
-  public void flush(Map<TopicPartition, OffsetAndMetadata> offsets) {
-    if (writer != null) {
-      log.debug("Flushing data to Elasticsearch with the following offsets: {}", offsets);
-      writer.flush();
-    } else {
-      log.debug("Could not flush data to Elasticsearch because ESWriter already closed.");
+  public String version() {
+    return Version.getVersion();
+  }
+
+  private void checkIndex(String index) {
+    if (!indexCache.contains(index)) {
+      log.info("Creating index {}.", index);
+      client.createIndex(index);
+      indexCache.add(index);
     }
   }
 
-  @Override
-  public void close(Collection<TopicPartition> partitions) {
-    log.debug("Closing the task for topic partitions: {}", partitions);
+  private void checkMapping(SinkRecord record) {
+    String index = convertTopicToIndexName(record.topic());
+    if (!config.shouldIgnoreSchema(record.topic()) && !existingMappings.contains(index)) {
+      if (!client.hasMapping(index)) {
+        client.createMapping(index, record.valueSchema());
+      }
+      log.debug("Locally caching mapping for index '{}'", index);
+      existingMappings.add(index);
+    }
   }
 
-  @Override
-  public void stop() throws ConnectException {
-    log.info("Stopping ElasticsearchSinkTask");
-    if (writer != null) {
-      writer.stop();
-      writer = null;
+  /**
+   * Returns the converted index name from a given topic name. Elasticsearch accepts:
+   * <ul>
+   *   <li>all lowercase</li>
+   *   <li>less than 256 bytes</li>
+   *   <li>does not start with - or _</li>
+   *   <li>is not . or ..</li>
+   * </ul>
+   * (<a href="https://github.com/elastic/elasticsearch/issues/29420">ref</a>_.)
+   */
+  private String convertTopicToIndexName(String topic) {
+    String index = topic.toLowerCase();
+    if (index.length() > 255) {
+      index = index.substring(0, 255);
     }
-    if (client != null) {
-      client.close();
+
+    if (index.startsWith("-") || index.startsWith("_")) {
+      index = index.substring(1);
     }
+
+    if (index.equals(".") || index.equals("..")) {
+      index = index.replace(".", "dot");
+      log.warn("Elasticsearch cannot have indices named {}. Index will be named {}.", topic, index);
+    }
+
+    log.trace("Topic '{}' was translated to index '{}'.", topic, index);
```

Comment:
Should it be `warn`?

Reply:
not necessarily, it could be a straight translation from the topic. I can add `warn` for every if branch instead if you like

Reply:
I think it's important information if we are changing the index name to be different from the expected name(topic name).